File: TheWatch-iOS/Logging/LogEntry.swift
import Foundation

/// Structured log entry — the domain model for all logging in TheWatch iOS.
///
/// Mirrors Serilog's structured logging semantics:
/// - `messageTemplate` with `{PropertyName}` placeholders
/// - `properties` dictionary for structured data
/// - `sourceContext` identifies the originating component
/// - `correlationId` links related events across an SOS lifecycle
struct LogEntry: Identifiable, Codable, Sendable {
    let id: String
    let timestamp: Date
    let level: LogLevel
    let sourceContext: String
    let messageTemplate: String
    let properties: [String: String]
    let exception: String?
    let correlationId: String?
    let userId: String?
    let deviceId: String?
    var synced: Bool

    init(
        id: String = UUID().uuidString,
        timestamp: Date = Date(),
        level: LogLevel,
        sourceContext: String,
        messageTemplate: String,
        properties: [String: String] = [:],
        exception: String? = nil,
        correlationId: String? = nil,
        userId: String? = nil,
        deviceId: String? = nil,
        synced: Bool = false
    ) {
        self.id = id
        self.timestamp = timestamp
        self.level = level
        self.sourceContext = sourceContext
        self.messageTemplate = messageTemplate
        self.properties = properties
        self.exception = exception
        self.correlationId = correlationId
        self.userId = userId
        self.deviceId = deviceId
        self.synced = synced
    }

    /// Render the message template with property values substituted.
    var renderedMessage: String {
        var result = messageTemplate
        for (key, value) in properties {
            result = result.replacingOccurrences(of: "{\(key)}", with: value)
        }
        return result
    }
}

File: TheWatch-Android/app/src/main/java/com/thewatch/app/data/logging/LogEntry.kt
package com.thewatch.app.data.logging

import java.time.Instant
import java.util.UUID

/**
 * Structured log entry — the domain model for all logging in TheWatch.
 *
 * Mirrors Serilog's structured logging semantics:
 * - messageTemplate with {PropertyName} placeholders
 * - properties map for structured data
 * - sourceContext identifies the originating component
 * - correlationId links related events across an SOS lifecycle
 *
 * @property id           Unique log entry identifier
 * @property timestamp    UTC instant the event occurred
 * @property level        Severity level (Verbose → Fatal)
 * @property sourceContext Fully-qualified class/component name (e.g. "LocationCoordinator")
 * @property messageTemplate Serilog-style template: "Location updated to {Lat},{Lng} in {Mode} mode"
 * @property properties   Structured key-value pairs extracted from the template + extras
 * @property exception    Optional exception message + stack trace
 * @property correlationId Optional SOS/incident ID linking related log entries
 * @property userId       Current authenticated user (null if pre-auth)
 * @property deviceId     Stable device identifier for cross-device correlation
 * @property synced       Whether this entry has been pushed to Firestore
 */
data class LogEntry(
    val id: String = UUID.randomUUID().toString(),
    val timestamp: Instant = Instant.now(),
    val level: LogLevel,
    val sourceContext: String,
    val messageTemplate: String,
    val properties: Map<String, String> = emptyMap(),
    val exception: String? = null,
    val correlationId: String? = null,
    val userId: String? = null,
    val deviceId: String? = null,
    val synced: Boolean = false
) {
    /** Render the message template with property values substituted. */
    fun renderedMessage(): String {
        var result = messageTemplate
        for ((key, value) in properties) {
            result = result.replace("{$key}", value)
        }
        return result
    }
}
